esp/src/src/Memory.ts

```diff
@@ -1,9 +1,71 @@
-import * as Deferred from "dojo/Deferred";
 import * as Observable from "dojo/store/Observable";
 import * as QueryResults from "dojo/store/util/QueryResults";
 import * as SimpleQueryEngine from "dojo/store/util/SimpleQueryEngine";
 import { alphanum } from "./Utility";
 
+class Deferred<T> {
+
+    promise: Promise<T>;
+    resolve: (value: T | PromiseLike<T>) => void;
+    reject: (reason?: any) => void;
+
+    protected _isCancelled = false;
+    protected _cancelledReason: string;
+    protected _isResolved = false;
+    protected _isRejected = false;
+
+    constructor() {
+        this.promise = new Promise((resolve, reject) => {
+            this.resolve = resolve;
+            this.reject = reject;
+        });
+    }
+
+    then(onFulfilled?: (value: T) => void, onRejected?: (reason: any) => void): Deferred<T> {
+        this.promise.then((value: T) => {
+            if (this._isCancelled && onRejected) {
+                onRejected(this._cancelledReason);
+            } else if (!this._isCancelled && onFulfilled) {
+                this._isResolved = true;
+                onFulfilled(value);
+            }
+        }, (reason: any) => {
+            if (this._isCancelled && onRejected) {
+                onRejected(this._cancelledReason);
+            } else if (!this._isCancelled && onRejected) {
+                this._isRejected = true;
+                onRejected(reason);
+            }
+        });
+        return this;
+    }
+
+    cancel(reason?: string) {
+        this._isCancelled = true;
+        this._cancelledReason = reason;
+    }
+
+    isResolved() {
+        return this._isResolved;
+    }
+
+    isRejected() {
+        this._isRejected;
+    }
+
+    isFulfilled() {
+        this._isResolved || this._isRejected || this._isCancelled;
+    }
+
+    isCanceled() {
```

Comment:
missing an L in the function name

Reply:
https://www.merriam-webster.com/words-at-play/canceled-or-cancelled

Reply:
The fact that it is spelled differently 3 lines higher is a bit of a shame...

Reply:
Fixed misspelling